### PR TITLE
Alit/mamba 2 0 migration

### PR DESCRIPTION
**Pull request by JRD971000**
# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/llm/gpt/data/fine_tuning.py [hunk 1]
```diff
@@ -67,6 +68,7 @@ def __init__(
         self.rampup_batch_size = rampup_batch_size
         self.data_sampler = None
         self.max_train_samples = None
+        self.pad_to_max_length = (pad_to_max_length,)
```
Comment (cuichenx) [inline nemo/collections/llm/gpt/data/fine_tuning.py:71]: Isn't `pad_to_max_length` a boolean?

File: nemo/collections/llm/gpt/model/ssm.py [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+
+
+class SSMModel(GPTModel):
+
+    def forward(self, input_ids, position_ids=None, attention_mask=None, labels=None):
+        attention_mask = None
+        output_tensor = self.module(
+            input_ids=input_ids, position_ids=position_ids, attention_mask=attention_mask, labels=labels
+        )
+        return output_tensor
```
Comment (cuichenx) [inline nemo/collections/llm/gpt/model/ssm.py:86]: Is it possible to override the `forward_step_fn` instead of subclassing?

File: nemo/collections/llm/__init__.py [hunk 3]
```diff
@@ -58,12 +63,16 @@
     Nemotron4Config340B,
     NemotronConfig,
     NemotronModel,
+    NVIDIAHybridConfig8b,
```
Comment (cuichenx) [inline nemo/collections/llm/__init__.py:66]: Can we add "Mamba" to this name, otherwise might be confusing out of context

File: nemo/collections/llm/gpt/model/ssm.py [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+from megatron.core import parallel_state
+from megatron.core.models.mamba import MambaModel as MCoreMambaModel
+from megatron.core.models.mamba.mamba_layer_specs import mamba_stack_spec
+from megatron.core.transformer.transformer_config import TransformerConfig
+
+from nemo.collections.llm.gpt.model.base import GPTModel, gpt_data_step, gpt_forward_step
+from nemo.lightning import get_vocab_size, io, teardown
+
+if TYPE_CHECKING:
+    from transformers import LlamaConfig as HFLlamaConfig
```
Comment (cuichenx) [inline nemo/collections/llm/gpt/model/ssm.py:16]: is this still needed?

**Comment by marcromeyn** [comment]
Thanks for this, looks good! My one comment is that we want to move away from scripts in the example-dir that we use inside GH-actions etc. For that, we would like to move to using a recipe (like we defined for many models here, a recipe should contain everything for training so model/data/trainer/etc.). 

We are currently working on automatically exposing those in the CLI, for more docs around that see [this](https://github.com/NVIDIA/NeMo/blob/e76afdfc28a1835fcf8aac97c07261e24ca0407d/examples/llm/pretrain/README.md) readme inside this [WIP PR](https://github.com/NVIDIA/NeMo/pull/10300).

So would be great if you could add one (or a few) recipes for this model, you could for instance take llama3_8b as an [example](https://github.com/NVIDIA/NeMo/blob/e76afdfc28a1835fcf8aac97c07261e24ca0407d/nemo/collections/llm/recipes/llama3_8b.py).

File: nemo/collections/llm/gpt/model/ssm.py (hunk 4 above)
Comment (JRD971000) [reply nemo/collections/llm/gpt/model/ssm.py:16]: removed

File: nemo/collections/llm/gpt/model/ssm.py (hunk 2 above)
Comment (JRD971000) [reply nemo/collections/llm/gpt/model/ssm.py:86]: very good point, I did this and it is much cleaner. Thanks!

File: nemo/collections/llm/gpt/data/fine_tuning.py (hunk 1 above)
Comment (JRD971000) [reply nemo/collections/llm/gpt/data/fine_tuning.py:71]: fixed

File: nemo/collections/llm/__init__.py (hunk 3 above)
Comment (JRD971000) [reply nemo/collections/llm/__init__.py:66]: changed to NVIDIAMambaHybridConfig8b

File: nemo/collections/llm/gpt/model/ssm.py [hunk 5]
```diff
@@ -0,0 +1,302 @@
+import re
+from dataclasses import dataclass
+from pathlib import Path
+from typing import TYPE_CHECKING, Annotated, Callable, Literal, Optional
```
[the remark this answers is not in the document]
Comment (JRD971000) [reply nemo/collections/llm/gpt/model/ssm.py:4]: fixed

File: nemo/collections/llm/gpt/model/ssm.py [hunk 6]
```diff
@@ -0,0 +1,302 @@
+import re
```
[the remark this answers is not in the document]
Comment (JRD971000) [reply nemo/collections/llm/gpt/model/ssm.py:1]: fixed

**Comment by JRD971000** [comment]
> Thanks for this, looks good! My one comment is that we want to move away from scripts in the example-dir that we use inside GH-actions etc. For that, we would like to move to using a recipe (like we defined for many models [here](https://github.com/NVIDIA/NeMo/tree/e76afdfc28a1835fcf8aac97c07261e24ca0407d/nemo/collections/llm/recipes), a recipe should contain everything for training so model/data/trainer/etc.).
> 
> We are currently working on automatically exposing those in the CLI, for more docs around that see [this](https://github.com/NVIDIA/NeMo/blob/e76afdfc28a1835fcf8aac97c07261e24ca0407d/examples/llm/pretrain/README.md) readme inside this [WIP PR](https://github.com/NVIDIA/NeMo/pull/10300).
> 
> So would be great if you could add one (or a few) recipes for this model, you could for instance take llama3_8b as an [example](https://github.com/NVIDIA/NeMo/blob/e76afdfc28a1835fcf8aac97c07261e24ca0407d/nemo/collections/llm/recipes/llama3_8b.py).

@marcromeyn Thanks for your comment, I will remove the script from the MR

File: nemo/collections/llm/gpt/model/ssm.py [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+    num_attention_heads: int = 32
+    num_query_groups: int = 8
+    make_vocab_size_divisible_by: int = 128
+    tokenizer_library: str = 'megatron'
+    tokenizer_name: str = "GPTSentencePieceTokenizer"
+    mapping_type: str = "nvidia-hybrid"
+
+
+__all__ = [
+    "SSMModel",
```
[the remark this answers is not in the document]
Comment (JRD971000) [reply nemo/collections/llm/gpt/model/ssm.py]: resolved

File: tests/collections/llm/megatron_ssm_finetuning.py [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+    data = llm.SquadDataModule(
+        seq_length=512,
+        micro_batch_size=2,
+        global_batch_size=4,
+        tokenizer=model.tokenizer,
+        num_workers=0,
+        pad_to_max_length=True,
+    )
+
+    app_state = _setup(
```
[the remark this answers is not in the document]
Comment (yaoyu-33) [reply tests/collections/llm/megatron_ssm_finetuning.py:97]: fix code scan

File: nemo/collections/llm/__init__.py [hunk 9]
```diff
@@ -31,6 +31,11 @@
     Baichuan2Config,
     Baichuan2Config7B,
     Baichuan2Model,
+    BaseMambaConfig1_3b,
```
Comment (yaoyu-33) [inline nemo/collections/llm/__init__.py:34]: change 3b to 3B to keep things consistent, similar below

File: nemo/collections/llm/__init__.py [hunk 10]
```diff
@@ -120,6 +128,14 @@
     "Nemotron4Config22B",
     "Nemotron4Config340B",
     "NemotronConfig",
+    "SSMConfig",
+    "BaseMambaConfig130m",
```
Comment (yaoyu-33) [inline nemo/collections/llm/__init__.py:132]: `m` -> `M`

File: nemo/collections/llm/gpt/model/ssm.py [hunk 11]
```diff
@@ -0,0 +1,294 @@
...
+    num_layers: int = 2
+    mamba_ssm_ngroups: int = 8
+    num_attention_heads: int = 1
+    hybrid_attention_ratio: float = 0.0
+    hybrid_mlp_ratio: float = 0.0
+    hybrid_override_pattern: str = None
+    post_process: bool = True
+    pre_process: bool = True
+    seq_length: int = 2048
+    params_dtype: torch.dtype = torch.bfloat16
```
Comment (yaoyu-33) [inline nemo/collections/llm/gpt/model/ssm.py:41]: where is this getting used?

**Review by ericharper — APPROVED** [review]
LGTM. Thanks!
